ext/semian_cb_data/semian_cb_data.c

```diff
@@ -0,0 +1,900 @@
+    return Qnil;
+
+  if (!semian_cb_data_lock(self))
+    return Qnil;
+
+  shared_cb_data *data = ptr->shm_address;
+  if (data->arr_length == ptr->arr_max_size) {
+    int i;
+    for (i=1; i< ptr->arr_max_size; ++i){
+      data->errors[i-1] = data->errors[i];
+    }
+    --(data->arr_length);
+  }
+  data->errors[(data->arr_length)] = NUM2DBL(num);
+  ++(data->arr_length);
+  semian_cb_data_unlock(self);
+  return self;
+}
+
+static VALUE
+semian_cb_data_set_pop_back(VALUE self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (0 == ptr->shm_address)
+    return Qnil;
+
+  if (!semian_cb_data_lock(self))
+    return Qnil;
+
+  VALUE retval;
+  shared_cb_data *data = ptr->shm_address;
+  if (0 == data->arr_length)
+    retval = Qnil;
+  else {
+    retval = DBL2NUM(data->errors[data->arr_length-1]);
+    --(data->arr_length);
+  }
+
+  semian_cb_data_unlock(self);
+  return retval;
+}
+
+static VALUE
+semian_cb_data_set_pop_front(VALUE self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (0 == ptr->shm_address)
+    return Qnil;
+
+  if (!semian_cb_data_lock(self))
+    return Qnil;
+
+  VALUE retval;
+  shared_cb_data *data = ptr->shm_address;
+  if (0 >= data->arr_length)
+    retval = Qnil;
+  else {
+    retval = DBL2NUM(data->errors[0]);
+    int i=0;
+    for (; i<data->arr_length-1; ++i)
+      data->errors[i]=data->errors[i+1];
+    --(data->arr_length);
+  }
+
+  semian_cb_data_unlock(self);
+  return retval;
+}
+
+static VALUE
+semian_cb_data_set_push_front(VALUE self, VALUE num)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (0 == ptr->shm_address)
+    return Qnil;
+  if (TYPE(num) != T_FIXNUM && TYPE(num) != T_FLOAT/*|| TYPE(size) != T_BIGNUM*/)
+    return Qnil;
+
+  if (!semian_cb_data_lock(self))
+    return Qnil;
+
+  double val = NUM2DBL(num);
+  shared_cb_data *data = ptr->shm_address;
+
+  int i=data->arr_length;
+  for (; i>0; --i)
+    data->errors[i]=data->errors[i-1];
+
+  data->errors[0] = val;
+  ++(data->arr_length);
+  if (data->arr_length>ptr->arr_max_size)
+    data->arr_length=ptr->arr_max_size;
+
+  semian_cb_data_unlock(self);
+  return self;
+}
+
+static VALUE
+semian_cb_data_array_clear(VALUE self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (0 == ptr->shm_address)
+    return Qnil;
+
+  if (!semian_cb_data_lock(self))
+    return Qnil;
+  ptr->shm_address->arr_length=0;
+
+  semian_cb_data_unlock(self);
+  return self;
+}
+
+static VALUE
+semian_cb_data_array_first(VALUE self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (0 == ptr->shm_address)
+    return Qnil;
+
+  if (!semian_cb_data_lock(self))
+    return Qnil;
+
+  VALUE retval;
+  if (ptr->shm_address->arr_length >=1 && 1 <= ptr->arr_max_size)
```

Comment:
Why would the `arr_max_size` ever not be 1? You already do this check when initializing, you don't need to do it again.